File: 2021/01/sonar_sweep.py
#!/usr/bin/env python3.10

from aoc2021 import *
from itertools import pairwise

p1(len([
    a for a, b in pairwise(
        int(l.strip()) for l in inp()
    ) if b > a
]))

def window(iter, n):
    for i in range(0, len(iter) - n + 1): yield iter[i:i+n]

p2(len([a for a, b, c, d in window(list(int(l.strip()) for l in inp()), 4) if (d) > (a)]))

File: 2021/01/sonar_sweep.rs
#!/usr/bin/env rustr

use aoc::*;

fn main() {
    let mut aoc = AdventOfCode::new(2021, 1);
    let inp = aoc.get_input();
    let measurements = inp.lines().map(|l| l.parse::<usize>().unwrap());

    let p1 = measurements
        .clone()
        .tuple_windows()
        .filter(|(a, b)| b > a)
        .count();
    aoc.submit_p1(p1).unwrap();

    let p2 = measurements
        .tuple_windows()
        .filter(|(a, _, _, d)| d > a)
        .count();
    aoc.submit_p2(p2).unwrap();
}
